### PR TITLE
RewardNet Tutorial

### LINKED ISSUE #714: Documentation for Reward Networks
## Problem
Right now there is no page in the docs describing how reward networks are used in `imitation`. There are only two tutorials and an article about density-based reward modeling.

## Solution
Write a page that describes how to create, train, load, save and use a reward network with various imitation algorithms.

## Related Issues
#698

### PR DESCRIPTION
**Pull request by levmckinney**
## Description
Adds basic documentation on reward networks. Resolves #714 and by extension #698.

## Testing
* Only documentation changes

## TODO
* [x] switch all totorials to using `predict_processed` instead of `predict` in when using the reward network to train a policy.
* [x] Update developer guid to point to new tutorial

### CODE REVIEW

File: docs/tutorials/reward_networks.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    from imitation.rewards.reward_nets import BasicRewardNet
+
+    reward_net = BasicRewardNet(obs_space, action_space)
+    venv = util.make_vec_env("Pendulum-v1", n_envs=3, rng=rng)
+    venv = RewardVecEnvWrapper(venv, reward_net.predict_processed)
+
+Reward Network Wrappers
+-----------------------
+
+Imitation learning algorithms should converge to a reward function that will theoretically induce the optimal/soft optimal policy. However, these reward functions may not always be well suited for training RL agents, or we may want to modify them to encourage exploration, for instance.
```
Comment (ernestum) [inline docs/tutorials/reward_networks.rst:68]: Maybe link to a definition of "soft optimal"?

File: docs/tutorials/reward_networks.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+There are two types of wrapper:
+* :class:`ForwardWrapper <imitation.rewards.reward_nets.ForwardWrapper>` allows for direct modification of the results of the reward network's ``forward`` method. It is used during the learning of the reward network and thus must be differentiable. These wrappers are always applied first and are always active. They are used for applying transformations like potential shaping (see :class:`ShapedRewardNet <imitating.rewards.reward_nets.ShapedRewardNet>`).
+* :class:`PredictProcessedWrapper <imitation.rewards.reward_nets.PredictProcessedWrapper>` modifies the predict_processed call to the reward network. Thus this type of reward network wrapper is designed to only modify the reward when it is being used to train/evaluate a policy but *not* when we are taking gradients on it. Thus it does not have to be differentiable.
+
```
Comment (ernestum) [inline docs/tutorials/reward_networks.rst:73]: Does not render correctly in sphinx. Try dashes and indentation an extra newline?

What doe you mean by "always active"? Why should they not be always active?

File: docs/tutorials/reward_networks.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    from imitation.rewards.reward_nets import NormalizedRewardNet
+    from imitation.util.networks import RunningNorm
+    train_reward_net = NormalizedRewardNet(
+        reward_net,
+        normalize_output_layer=RunningNorm,
+    )
+
+.. note::
+    The reward normalization wrapper does _not_ function identically to stable baselines3's `VecNormalize <https://stable-baselines3.readthedocs.io/en/master/guide/vec_envs.html#stable_baselines3.common.vec_env.VecNormalize>`_ environment wrapper. First, it does not normalize the observations. Second it normalizes the reward based on the reward networks's mean and variance on and *not* a running estimate of the return.
+
```
Comment (ernestum) [inline docs/tutorials/reward_networks.rst:88]: This note confuses me. You seem to say that the NromalizedRwardNet Wrapper on a running estimate but then what you describe below *does* sound like a running estimate.

Also I think the last sentence might have one "on" too many.

File: docs/tutorials/reward_networks.rst [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+.. testcode::
+
+    from functools import partial
+    eval_rew_fn = partial(reward_net.predict_processed, update_stats=False)
+
+Serializing and Deserializing Reward Networks
+---------------------------------------------
+
+Reward networks are serialized simply by calling ``th.save(reward_net, path)``. When evaluating reward networks, we may or may not want to include the wrappers it was trained with. To load the raw reward network, we can simply call ``th.load(path)``.
+
```
Comment (ernestum) [inline docs/tutorials/reward_networks.rst:100]: On the first pass reading it was not clear here whether `th.save(reward_net, path)` will also save the wrappers and if `th.load(path)` would restore them. I then read the code to clarify and came back to the description to read it a second time. Then it was clear.

Maybe you can help (quick) understanding by adding to the first sentence:

"Reward networks and **their wrappers** are serialized simply by calling ..."

and 

"to laod the raw reward network **without wrappers applied**, we can simply call ..."

**Review by ernestum — CHANGES_REQUESTED** [review]
This is clearly written and after reading it I feel that I got a good understanding about thow the reward networks are used and what the design choices were when implementing them.

File: src/imitation/rewards/reward_nets.py [hunk 5]
```diff
@@ -149,7 +149,8 @@ def predict_th(
             with th.no_grad():
                 rew_th = self(state_th, action_th, next_state_th, done_th)
 
-            assert rew_th.shape == state.shape[:1]
+            assert rew_th.shape == state.shape[:1] # TODO do we or don't we support 
+            # state spaces with more than one axis?
```
[the remark this answers is not in the document]
Comment (levmckinney) [reply src/imitation/rewards/reward_nets.py]: I'm just being dumb and mis read this code its fine.

**Pull request merged**